GenTs.cs

```diff
@@ -0,0 +1,297 @@
+        case CiCallType.Virtual:
+          break;
+        case CiCallType.Abstract:
+          Write("abstract ");
+          break;
+        case CiCallType.Override:
+          break;
+        case CiCallType.Normal:
+          // no final methods in TS
+          break;
+        case CiCallType.Sealed:
+          // no final methods in TS
+          break;
+        default:
+          throw new NotImplementedException(method.CallType.ToString());
+      }
+      WriteName(method);
+      Write('(');
+      int i = 0;
+      foreach (CiVar param in method.Parameters)
+      {
+        if (i >= paramCount)
+          break;
+        if (i > 0)
+          Write(", ");
+        WriteName(param);
+        if (param.Value != null)
+          Write('?');
+        Write(": ");
+        Write(param.Type);
+        i++;
+      }
+      Write("): ");
+      Write(method.Type);
+      WriteLine(";");
+    }
+
+    void WriteConsts(IEnumerable<CiConst> consts)
+    {
+      foreach (CiConst konst in consts)
+      {
+        WriteLine();
+        Write(konst.Documentation);
+        Write(konst.Visibility);
+        Write("static readonly ");
+        WriteTypeAndName(konst);
+        WriteLine(';');
+      }
+    }
+
+    void Write(CiClass klass)
+    {
+      Write(klass.Documentation);
+      Write("export ");
+      switch (klass.CallType)
+      {
+        case CiCallType.Normal:
+          break;
+        case CiCallType.Abstract:
+          Write("abstract ");
+          break;
+        case CiCallType.Static:
+        case CiCallType.Sealed:
+          // there's no final/sealed keyword, but we can accomplish it by marking the constructor private
+          break;
+        default:
+          throw new NotImplementedException(klass.CallType.ToString());
+      }
+      OpenClass(klass, "", " extends ");
+
+      if (klass.CallType == CiCallType.Static || klass.CallType == CiCallType.Sealed)
+      {
+        Write("private constructor();");
+      }
+      else if (klass.Constructor != null && klass.Constructor.Visibility != CiVisibility.Public)
+      {
+        if (klass.Constructor != null)
+        {
+          Write(klass.Constructor.Documentation);
+          Write(klass.Constructor.Visibility);
+        }
+        WriteLine("constructor();");
+      }
+
+      WriteConsts(klass.Consts);
+
+      foreach (CiField field in klass.Fields)
+      {
+        Write(field.Visibility);
+        WriteTypeAndName(field);
+        WriteLine(';');
+      }
+
+      foreach (CiMethod method in klass.Methods)
+      {
+        WriteSignature(method, method.Parameters.Count);
+      }
+
+      WriteConsts(klass.ConstArrays);
+      CloseBlock();
+      WriteLine();
+    }
+
+
+    public override void Write(CiProgram program)
+    {
+      CreateFile(this.OutputFile);
+      foreach (CiEnum enu in program.OfType<CiEnum>())
+        Write(enu);
+      foreach (CiClass klass in program.OfType<CiClass>()) // TODO: topological sort of class hierarchy
```

Comment:
Regarding the `TODO` comment: does TypeScript support forward references? e.g.

```csharp
class Foo
{
    public void foo(Bar b);
}

class Bar
{
}
```

BTW. I'm not sure if that's a problem for JS (think object storage, not pointers).

Reply:
There's no problem with forward references of any kind in type annotations (btw the syntax is `public foo(b: Bar): void;`).

As far as forward references at runtime, function declarations are ["hoisted"](https://www.digitalocean.com/community/tutorials/understanding-hoisting-in-javascript#hoisting-functions) to the start of the containing block so forward references won't cause any runtime errors.  This is not to be confused with assigning a variable to a function *expression* however, which isn't hoisted (the link explains this)

Reply:
And JS classes are essentially just functions so they can be forward referenced as well

Reply:
If static property initializers refer to static properties of other classes though, that could cause problems, so a toposort would be needed.

Reply:
What do you mean by static properties?

Reply:
Static fields of the class, for example

```ts
class RECOIL
{
    readonly VersionMajor: number = 5;
}
```

There's a proposal to support them in ECMAScript (without readonly modifier) as well: https://github.com/tc39/proposal-class-fields

Reply:
Constants are resolved by `cito`, so the generator is expected to emit values as literals, not references.

What I'm worried about is inheritance. In JS this is emitted as:
```js
DerivedClass.prototype = new BaseClass();
```
BTW. Perhaps JS should switch to the new `class` syntax? What's the adoption by the browsers and server-side technologies? I already emit `let`/`const` instead of `var`.